broker/ga4gh/broker/endpoints/builds.py

```diff
@@ -0,0 +1,504 @@
+import datetime
+import logging
+import os
+import shutil
+from typing import (Dict)
+
+import yaml
+from flask import (current_app)
+from git import Repo
+from git.exc import GitCommandError
+from kubernetes import client, config
+from kubernetes.client import ApiException
+from pymongo.errors import DuplicateKeyError
+from werkzeug.exceptions import Unauthorized
+
+from broker.errors.exceptions import (RepositoryNotFound,
+                                      BuildNotFound, DeletePodError,
+                                      CreatePodError, WrongGitCommand)
+from broker.ga4gh.broker.endpoints.repositories import generate_id
+from broker.ga4gh.broker.endpoints.subscriptions import notify_subscriptions
+
+logger = logging.getLogger(__name__)
+
+template_file = '/app/broker/ga4gh/broker/endpoints/template/template.yaml'
+
+
+def register_builds(repository_id: str, access_token: str, build_data: Dict):
+    """Register new builds for already registered repository.
+
+    Args:
+        repository_id: Identifier for repository.
+        access_token: Secret used to verify source of the request to
+        initiate new build.
+        build_data: Request data containing build information.
+
+    Returns:
+        build_id: Identifier for new registered build.
+
+    Raises:
+        Unauthorized: Raised when access_token is invalid or not specified
+        in request.
+        RepositoryNotFound: Raised when repository is not found with given
+        identifier.
+
+    Description:
+        - Takes request build_data (Build information).
+        - Check if repository with specified identifier is available or not.
+        - Verify access_token.
+        - Generate a random string, add it after repository identifier and
+        use it as build identifier. (Retries 3 times for generating unique
+        random string.)
+        - Insert build_data in mongodb.
+        - Call create_build function.
+        - returns build_id.
+    """
+    retries = 3
```

Comment:
Just a suggestion here: It is better to define a separate class with dedicated methods to do specific functions. That class instance can then be called inside the controller.

For example, this can be further divided into small functionalities:
* Check and verify input data, if required (repo not found)
* If found, perform operations: one function to return a build payload, and another to create the build
Check how we have implemented the Tool Registry classes under TRS-Filer.

Great work on modularising and clubbing builds, repositories, and dedicated methods into separate dedicated files, just that I prefer class-based implementations slightly more.